### PR TITLE
ENH: New boundary inputs

### LINKED ISSUE #40245: ENH: consistency of input args for boundaries
There are some methods where boundary inputs are required.

This issue proposes standardising to the keyword arg: `inclusive` and using values `{"both", "neither", "left", "right"}`.

- [x] Styler.highlight_between():  `'inclusive' in {"both", "neither", "left", "right"}` 
- [x] Series.between() (#40628): `'inclusive' in {"both", "neither", "left", "right"}` 
- [x] pd.date_range() https://github.com/pandas-dev/pandas/pull/43504
- [x] DataFrame.between_time() https://github.com/pandas-dev/pandas/pull/43248
- [ ] pd.Interval https://github.com/pandas-dev/pandas/pull/46522
- [ ] pd.interval_range https://github.com/pandas-dev/pandas/pull/46355
- [ ] pd.IntervalIndex https://github.com/pandas-dev/pandas/pull/46522
- [ ] pd.cut
- [ ] Series.resample
- [ ] DataFrame.rolling(closed)

More methods yet to be identified.

### PR DESCRIPTION
**Pull request by hewittk**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] tests added / passed
- [x] closes #40245
- [x] whatsnew entry

### CODE REVIEW

**Comment by hewittk** [comment]
This pull request makes for more consistency and flexibility of the arguments for series.between and the validate_endpoints method of date_timelike.py which is the underlying method of `closed` for date_range. The parameters that can be inputted to the methods are either a boolean value or a string value of 'both', 'left', 'right', or 'neither'.

**Comment by mzeitlin11** [comment]
Thanks for the pr @hewittk! Looks like some unrelated changes crept in - seeing some `pip` related additions which should be reverted

**Comment by hewittk** [comment]
Thank you for letting me know @mzeitlin11! The pip files should be deleted now, let me know if there are any additions still there that I haven't noticed.

**Comment by mzeitlin11** [comment]
> Thank you for letting me know @mzeitlin11! The pip files should be deleted now, let me know if there are any unrelated additions still there that I haven't noticed.

Looks good now. I think codecov is failing - if you added new options to parameters, we need to make sure that they're all tested (and a whatsnew note should be added explaining user-facing changes there are here)

File: pandas/core/arrays/datetimelike.py
Comment (attack68) [inline pandas/core/arrays/datetimelike.py:1819]: If we move towards standardisation is it better to document the argument inputs as `{"neither", "both", "left", "right"}` including some notes that `True` `False` and `None` remain accepted for backwards compatability?

File: pandas/core/arrays/datetimelike.py
Comment (attack68) [inline pandas/core/arrays/datetimelike.py:1833]: I would use the standard value first, i.e. `if closed == "neither" or (# additional conditionals for backwards compat)`.

Also I think you mean `closed=="both"` here and not `closed="neither"`.

File: pandas/core/arrays/datetimelike.py
Comment (attack68) [inline pandas/core/arrays/datetimelike.py:1840]: I think this is `closed="neither"`

File: pandas/core/series.py
Comment (attack68) [inline pandas/core/series.py:4687]: As before do the docs become: `inclusive : {"both", "neither", "left", "right"}`

where True and False are silently accepted.

File: pandas/core/series.py
Comment (attack68) [inline pandas/core/series.py:4741]: standard value first

File: pandas/core/series.py
Comment (attack68) [inline pandas/core/series.py:4750]: standard first

File: pandas/core/series.py
Comment (attack68) [inline pandas/core/series.py:4756]: Should we exclude the boolean if it is being silently hidden? Also this needs tests for code coverage of the raised error.

**Review by attack68 — CHANGES_REQUESTED** [review]
Thanks @hewittk for the PR. I made some comments inline. Some of the changes minght be considered non-controversial, and others we might some input from pandas core devs.

File: pandas/core/series.py
Comment (hewittk) [reply pandas/core/series.py]: How do you suggest that we write this line of documentation if inclusive has a standard value of a string ("both"), but also allows boolean values for backward compatibility?

File: pandas/core/arrays/datetimelike.py
Comment (hewittk) [reply pandas/core/arrays/datetimelike.py]: That does sound better, but where do you recommend putting the notes that `True`, `False`, and `None` remain accepted?

File: pandas/core/series.py
Comment (hewittk) [reply pandas/core/series.py]: That's a good question about hiding the boolean, right now I have it so that the standard string values are suggested first and then the boolean is mentioned. Do you think it would be best to exclude the boolean being the arrangement that we have right now?
We're working on getting more documented tests for code coverage over these next few days and should have wider coverage for our additions soon.

File: Dockerfile
Comment (attack68) [inline Dockerfile:5]: revert this

File: doc/source/whatsnew/v1.3.0.rst
Comment (attack68) [inline doc/source/whatsnew/v1.3.0.rst:144]: think you have picked these up from another PR, these are not related to your PR, I think so need to revert.

File: doc/source/whatsnew/v1.3.0.rst
Comment (attack68) [inline doc/source/whatsnew/v1.3.0.rst:145]:
```suggestion
- Add consistency of arguments in :meth:`Series.between` and :meth:`.datetimelike.validate_endpoints` to have ``{"left", "right", "neither", "both"}`` as standard argument values for both methods (:issue:`40628`)
```

File: pandas/core/arrays/datetimelike.py
Comment (attack68) [inline pandas/core/arrays/datetimelike.py:1816]:
```suggestion
    Check that the ``closed`` argument is a valid string input, or, 
    for backwards compatibility, one of ``{True, False, or None}``.
```

File: pandas/tests/series/methods/test_between.py
Comment (attack68) [inline pandas/tests/series/methods/test_between.py:43]: add GH issue number to added test.

To 'lock in' your changes best to ensure tests exist for:

  - [ ] accepting all (inc new) arguments in `validate_endpoints` as well as old ones for backwards compat, probably as two separate tests: and correct boundary functionality.
  - [ ] accepting all (inc new) arguments in `Series.between` probably as as two sep tests, and correct boundary functionality.
  - [ ] making sure the errors are hit with bad input

File: pandas/tests/series/methods/test_between.py
Comment (attack68) [inline pandas/tests/series/methods/test_between.py:46]: try to write a multiline str instead of combining two single lines

**Review by attack68 — CHANGES_REQUESTED** [review]
nice changes, but a few more I've noticed, comments inline.

Also which functions call `validate_endpoints` this seems to be a helper function designed to work with other functions with boundary options. If we change it then surely it will have an impact on the user api functions that call it?

File: pandas/core/series.py
Comment (attack68) [reply pandas/core/series.py]: I would probably do something like:

```
include : {"both", "neither", "left", "right"}
    Whether to set each bound as closed or open. For backwards compatibility this can also 
    be set to ``True`` ("both"), False ("neither") or None (..) 
```

You might find this is a bit more controversial with the core devs but I think it makes it quite evident that the users is supposed to adopt the recommended strings.

File: pandas/tests/series/methods/test_between.py
Comment (declanjcasey) [reply pandas/tests/series/methods/test_between.py:43]: @attack68 Thank you for the feedback, how exactly do I go about adding an issue number to the test case? Also should there be an issue number for each new test added?

File: pandas/tests/series/methods/test_between.py
Comment (declanjcasey) [reply pandas/tests/series/methods/test_between.py:46]: @attack68 thanks for your feedback! I tried using that originally but was thrown an error that it didn't match the regex as there's technically a `/n`. When using just a one-line variable declaration there is an error thrown that there's two many characters in one line. This is the only way I have found it to work so far.

Comment (attack68) [reply pandas/tests/series/methods/test_between.py:46]: https://stackoverflow.com/a/10660443

**Comment by attack68** [comment]
@hewittk I just took a look at what `validate_endpoints` feeds and there are quite a few time-based functions that use it and have `closed` as a boundary option, so there will likely be lots more changes needed for everything to go through.

Instead, and since this is your first PR, I would recommend you make only changes to `Series.between` and add the relevant tests and documentation changes just for that method, since it is less controversial, better defined and easier for devs to review. Of course after that it gives a basis for attempting further standardisation in other functions.

File: doc/source/whatsnew/v1.3.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:145]: can you leave the existing and only add a new note

Comment (jreback) [reply doc/source/whatsnew/v1.3.0.rst:145]: also no need to mention a private method

File: pandas/core/arrays/datetimelike.py
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:1866]: I guess we should deprecat this. can you separate this case out (for True/False/None) and show a FutureWarning (may have quite a number of deprecating warnings in the tests so need to change tests)

File: pandas/core/series.py
Comment (jreback) [inline pandas/core/series.py:4749]: can you add a versionchanged 1.3 here

File: doc/source/whatsnew/v1.3.0.rst
Comment (jreback) [reply doc/source/whatsnew/v1.3.0.rst:145]: move this to the deprecation section

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by hewittk** [comment]
> This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.

I apologize for the lack of activity, I have been very busy with midterms lately. I will work towards closing the request early this weekend.

**Comment by hewittk** [comment]
I believe that I have made all suggested changes now. I decided to go with @attack68's idea to only make changes to `Series.between` because its implementation and definition are simpler due to amount of functions that `validate_endpoints` underlies. 
The continuous integration tests don't appear to be running on my contributions to this branch now and I'm not sure why that is. There also appears to be a conflict in the whatsnew file even though I wasn't aware of myself removing any other entries from the depreciation section and there aren't removals to the file showing in the files changed.

**Comment by lithomas1** [comment]
@hewittk Can you resolve conflicts? I think CI is not triggered if conflicts are present.

File: pandas/core/series.py
Comment (jreback) [inline pandas/core/series.py:5006]: handle the bool values (the deprecated ones) as separate cases to emit a warning (and then simply translate them the correct inclusive)

File: pandas/core/series.py
Comment (jreback) [inline pandas/core/series.py:5019]: does a test hit this?

File: pandas/core/series.py
Comment (jreback) [inline pandas/core/series.py:4950]: don't say backwards compatibile. instead deprecated.

File: pandas/tests/series/methods/test_between.py
Comment (jreback) [inline pandas/tests/series/methods/test_between.py:43]: put the issue number a s a comment *inside* the test

File: pandas/tests/series/methods/test_between.py
Comment (jreback) [inline pandas/tests/series/methods/test_between.py:72]: make this a separate test is ok (call it test_between_error_args) or similar

**Review by jreback — CHANGES_REQUESTED** [review]
we need to add some tests for the deprecated args & fix any current tests to use the new args or use assert_produces_warning to catch the deprecation message

**Review by attack68 — CHANGES_REQUESTED** [review]
tried to add some code suggestions of my own, and also after jreback suggestions.

File: pandas/core/arrays/datetimelike.py
Comment (hewittk) [reply pandas/core/arrays/datetimelike.py]: I didn't create it, it looks like it was accidentally carried over from something else. I've deleted it now.

File: pandas/tests/series/methods/test_between.py
Comment (attack68) [inline pandas/tests/series/methods/test_between.py:69]:
```suggestion
```

should remove these since they are being tested below explicitly for deprecation warning

Comment (attack68) [reply pandas/tests/series/methods/test_between.py:69]: actually maybe put them in the test `test_between_inclusive_warning` 

The reason is that when someone actually removes these possibilities on a deprecation cycle this test is unaffected, whereas `test_between_inclusive _warning` can simply be deleted.

File: pandas/tests/series/methods/test_between.py
Comment (attack68) [inline pandas/tests/series/methods/test_between.py:43]: rename `test_between_inclusive_string` if the booleans removed as advised.

**Review by attack68 — CHANGES_REQUESTED** [review]
looks good except for the minor changes.

hopefully this gets in in time for 1.3.0 release, otherwise we might have to amend the version numbers.

File: pandas/core/series.py
Comment (simonjayhawkins) [inline pandas/core/series.py:5065]: I think maybe better to have `if inclusive == "both" or inclusive is True`  (and maybe keep the default as True) so that we ensure that the this works as before until the deprecation is actually enforced. (and the same for 'inclusive=False`)

File: pandas/core/series.py
Comment (simonjayhawkins) [inline pandas/core/series.py:5072]: can we deduplicate this warning code.

File: doc/source/whatsnew/v1.3.0.rst
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.3.0.rst:682]: maybe 2 release notes, one for the deprecation and another describing the new functionality.

File: pandas/core/series.py
Comment (simonjayhawkins) [reply pandas/core/series.py:5065]: I see @jreback already commented on this https://github.com/pandas-dev/pandas/pull/40628#discussion_r629419084

can you move the True/False cases to be first and translate the values.

File: pandas/core/series.py
Comment (simonjayhawkins) [reply pandas/core/series.py:5072]: maybe something like...
``` 
        if inclusive is True or inclusive is False:
            warnings.warn(
                "Boolean inputs to the `inclusive` argument are deprecated in"
                "favour of `both` or `neither`.",
                FutureWarning,
                stacklevel=2,
            )
       inclusive = "both" if inclusive else "neither"
```

and then the logic you have now. On master the condition is actually based on the truthiness of `inclusive` so maybe need to maintain that behavior but the docstring clearly states `bool` so not strictly necessary.

File: pandas/core/series.py
Comment (hewittk) [reply pandas/core/series.py]: Alright I see what you guys mean now; I misunderstood the previous comment. I changed it so that the True/False cases are handled and translated first.

File: pandas/core/series.py
Comment (hewittk) [reply pandas/core/series.py:4993]: What do you mean by a deprecated tag? 
I originally kept the bool argument but it was reviewed by @attack68 [here](https://github.com/pandas-dev/pandas/pull/40628#discussion_r602051448). So I'm not sure what the best practice for it in this case should be.

Comment (attack68) [reply pandas/core/series.py:4991]: The deprecation warning that appears **if** a boolean is used for these arguments should be enough, actually. 

We can de-pollute the docstring and remove any references to boolean. Users going forward should use the string values and only see that there are string values to use.

Comment (simonjayhawkins) [reply pandas/core/series.py:4991]: I could be wrong, but I'm pretty sure we use the `.. deprecated:: 1.3.0` elsewhere and not just remove stuff.

Comment (attack68) [reply pandas/core/series.py:4991]: @simonjayhawkins you could very well be right, but previously I've only seen deprecation tag referring to a deprecated method, or possibly a deprecated argument, but not to an input of an argument that is maintained?

**Review by attack68 — APPROVED** [review]
@hewittk thanks I think looks good to go in, and is in a good state to remove the future warning a few versions later to satisfy the full deprecation.

I'm not sure if this will make the time cut for 1.3.0 - we can wait and see, otherwise will have to shift to 1.4.0.

File: doc/source/whatsnew/v1.3.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:273]: add the issue number (you can use the pr number if you want here instead)

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment. pls merge master and ping when green.

File: doc/source/whatsnew/v1.3.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:719]: hmm you have a rebase issue. only *your* changes should show here (this section was moved so think adding it back from where it was)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @hewittk docstring validation failure to be resolved before merge, otherwise lgtm

**Comment by jreback** [comment]
@hewittk can you merge master and fix the doc-string

**Comment by jreback** [comment]
there is a test failure on this build: https://github.com/pandas-dev/pandas/pull/40628/checks?check_run_id=2909716187
likely need to catch the FutureWarning

**Review by simonjayhawkins — APPROVED** [review]

**Comment by simonjayhawkins** [comment]
Thanks @hewittk

**Comment by jbrockmendel** [comment]
@hewittk want to make a PR enforcing this deprecation?

**Pull request merged**
